File: python/valuecell/server/api/schemas/stock_analysis_research_task.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class StockAnalysisResearchTaskItemData(BaseModel):
    task_id: int
    thread_id: int
    user_id: str
    title: str
    summary: str
    task_type: str
    status: str
    priority: str
    source_kind: str
    source_ref: str | None = None
    related_tickers_json: list[str] = Field(default_factory=list)
    related_themes_json: list[str] = Field(default_factory=list)
    related_context_ids_json: list[int] = Field(default_factory=list)
    related_memory_id: int | None = None
    related_compression_id: int | None = None
    related_message_id: str | None = None
    resolution_note: str | None = None
    dismiss_reason: str | None = None
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None = None
    dismissed_at: datetime | None = None


class StockAnalysisResearchTaskListData(BaseModel):
    thread_id: int
    items: list[StockAnalysisResearchTaskItemData] = Field(default_factory=list)
    count: int = 0
    open_count: int = 0
    high_priority_open_count: int = 0
    last_generated_at: datetime | None = None
    has_actionable_gap: bool = False
    actionable_gap_summary: str | None = None
    generated_at: datetime


class StockAnalysisResearchTaskCreateRequest(BaseModel):
    title: str
    summary: str = ""
    task_type: str = "next_question"
    priority: str = "medium"
    source_kind: str = "manual"
    source_ref: str | None = None
    related_tickers_json: list[str] = Field(default_factory=list)
    related_themes_json: list[str] = Field(default_factory=list)
    related_context_ids_json: list[int] = Field(default_factory=list)
    related_memory_id: int | None = None
    related_compression_id: int | None = None
    related_message_id: str | None = None


class StockAnalysisResearchTaskGenerateRequest(BaseModel):
    pass


class StockAnalysisResearchTaskGenerateData(BaseModel):
    thread_id: int
    created_count: int = 0
    updated_count: int = 0
    items: list[StockAnalysisResearchTaskItemData] = Field(default_factory=list)
    last_generated_at: datetime | None = None
    summary: str | None = None


class StockAnalysisResearchTaskStateRequest(BaseModel):
    note: str | None = None


class StockAnalysisResearchTaskMutationData(BaseModel):
    thread_id: int
    task: StockAnalysisResearchTaskItemData

File: frontend/src/types/stock-analysis-research-task.ts
export type StockAnalysisResearchTask = {
  task_id: number;
  thread_id: number;
  user_id: string;
  title: string;
  summary: string;
  task_type:
    | "compare_followup"
    | "refresh_needed"
    | "tooling_check"
    | "memory_recheck"
    | "compression_recheck"
    | "risk_recheck"
    | "next_question"
    | "thesis_validation";
  status: "open" | "completed" | "dismissed";
  priority: "high" | "medium" | "low";
  source_kind:
    | "manual"
    | "memory"
    | "compression"
    | "compare"
    | "refresh"
    | "tooling"
    | "assistant_suggestion"
    | string;
  source_ref?: string | null;
  related_tickers_json: string[];
  related_themes_json: string[];
  related_context_ids_json: number[];
  related_memory_id?: number | null;
  related_compression_id?: number | null;
  related_message_id?: string | null;
  resolution_note?: string | null;
  dismiss_reason?: string | null;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
  dismissed_at?: string | null;
  is_focus_related?: boolean;
  is_related_to_latest_message?: boolean;
  is_planning_influencer?: boolean;
  is_primary_research_anchor?: boolean;
  suggested_planning_action?: string;
};

export type StockAnalysisResearchTaskList = {
  thread_id: number;
  items: StockAnalysisResearchTask[];
  count: number;
  open_count: number;
  high_priority_open_count: number;
  last_generated_at?: string | null;
  has_actionable_gap: boolean;
  actionable_gap_summary?: string | null;
  generated_at: string;
};

export type StockAnalysisResearchTaskMutationResult = {
  thread_id: number;
  task: StockAnalysisResearchTask;
};

export type StockAnalysisResearchTaskGenerateResult = {
  thread_id: number;
  created_count: number;
  updated_count: number;
  items: StockAnalysisResearchTask[];
  last_generated_at?: string | null;
  summary?: string | null;
};
